java/algorithm/pom.xml

```diff
@@ -48,5 +52,87 @@
     </dependency>
   </dependencies>
 
-  <build></build>
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-checkstyle-plugin</artifactId>
+        <version>3.1.0</version>
+        <configuration>
+          <skip>true</skip>
+        </configuration>
+      </plugin>
+    </plugins>
+  </build>
+
+  <profiles>
+    <profile>
+      <id>spotless</id>
+      <activation>
+        <activeByDefault>true</activeByDefault>
+      </activation>
+      <build>
+        <plugins>
+          <plugin>
+            <groupId>com.diffplug.spotless</groupId>
+            <artifactId>spotless-maven-plugin</artifactId>
+            <version>${spotless.version}</version>
+            <configuration>
+              <formats>
+                <format>
```

Comment:
This can be combined with the `pom` section below

Reply:
`for parameter formats: Cannot find 'formats' in class com.diffplug.spotless.maven.pom.Pom` I am getting this :thinking:

Reply:
Sorry, I should have been clearer. What I meant is that you could use those instead:

```
           <java>
             <googleJavaFormat>
               <version>1.7</version>
               <style>GOOGLE</style>
             </googleJavaFormat>
             <licenseHeader>
               <file>../dev/license/asf-java.license</file>
               <delimiter>package</delimiter>
             </licenseHeader>
           </java>
```

and
 
```
           <pom>
             <sortPom></sortPom>
             <licenseHeader>
               <file>../dev/license/asf-xml.license</file>
               <delimiter>(&lt;configuration|&lt;project)</delimiter>
             </licenseHeader>
           </pom>
```

That said, I also noticed by checking the change locally that the path to the license file is relative to the root of the module, which may cause extra issues (something that `maven-checkstyle-plugin` does support although not quite [documented](https://issues.apache.org/jira/browse/MCHECKSTYLE-131)). We may need to use `${maven.multiModuleProjectDirectory}` property (although not offically supported) or a plugin to compute the right path

Reply:
Thanks @laurentgo for the detailed response. I will try this out.

Reply:
The configuration for pom could be moved to the parent with path being `${maven.multiModuleProjectDirectory}/java/dev/license/asf-xml.license}` (should be safe to use because the project has a `.mvn` directory present). As it will reformat several existing `pom.xml` who inherits the configuration (there's a small indentation change), maybe we should address it as a standalone change like we did for #41173

As for the java format, yes it would be ideal to move to the parent as well but there's no good mechanism yet to limit the amount of change. The best I can think of is to do this in the parent pom

```
  <properties>
    <!-- replace with empty string to enable java formatting in submodules -->
    <spotless.java.excludes>**</spotless.java.excludes>
  </properties>

  <build>
    <plugins>
      <plugin>
         <groupId>com.diffplug.spotless</groupId>
         <artifactId>spotless-maven-plugin</artifactId>
         <configuration>
           <java>
             <excludes>
               <exclude>${spotless.java.excludes}</exclude>
             </excludes>
             <googleJavaFormat>
               <version>1.7</version>
               <style>GOOGLE</style>
             </googleJavaFormat>
             <licenseHeader>
               <file>${maven.multiModuleProjectDirectoryjava/dev/license/asf-java.license</file>
               <delimiter>package</delimiter>
             </licenseHeader>
           </java>
           [...]
         </configuration>
         [...]
      </plugin>
    </plugins>
  </build>
```

This would limit the amount of changes to the other pom.xml files and the final cleanup to be done by a large amount

Reply:
> The configuration for pom could be moved to the parent with path being ${maven.multiModuleProjectDirectory}/java/dev/license/asf-xml.license} (should be safe to use because the project has a .mvn directory present). As it will reformat several existing pom.xml who inherits the configuration (there's a small indentation change), maybe we should address it as a standalone change like we did for https://github.com/apache/arrow/issues/41173

Yes, this what I am also suggesting. We should do it in a single PR later on. And when we do this change per module, the diff on the last PR would be smaller or it could be just the config.

Reply:
> As for the java format, yes it would be ideal to move to the parent as well but there's no good mechanism yet to limit the amount of change. The best I can think of is to do this in the parent pom

Yes I think we can eventually move it in a separate PR as the last PR to clean things up. Because all the other PRs made per module would already have formatted the Java files per module.

Reply:
Overall I guess I was trying to advocate for two things:
* remove the license format for maven `pom.xml` from this review and instead do a PR to address all modules at once
* for java reformat, try to find a solution which require the less amount of lines to be added to each module

Reply:
> remove the license format for maven pom.xml from this review and instead do a PR to address all modules at once

I think we can do that. 

> for java reformat, try to find a solution which require the less amount of lines to be added to each module

Agreed. But we may not be able to control that limit. But I hear you. 

Thanks for the suggestions @laurentgo I will update accordingly.

Reply:
@laurentgo `${maven.multiModuleProjectDirectory}` property doesn't work for me locally? Do I have to enable something?

Reply:
No, you should not have. As long as you're working on a copy of the repo with the `.mvn` directory present at the root and maven 3.3.1 or higher

You can check the value by using this commandline inside the arrow  repo:
```
mvn -q -N help:evaluate -Dexpression=maven.multiModuleProjectDirectory -DforceStdout 
```

Reply:
It shows the expected path. Let me try this again.

Reply:
Ah it is working now. Must have been a glitch in my build.

Reply:
CIs are failing, seems the path is wrong.

Reply:
Change for #41429 moved java build from out of tree by copying `java` directory without also copying `.mvn` directory. This causes `${maven.multiModuleProjectDirectory} to return  the equivalent of `arrow/java` instead of `arrow`. It also means that the develocity extension is not configured properly because of it

Reply:
This is still an open issue right? I am not sure I can find a related PR?

Reply:
This [change](https://github.com/laurentgo/arrow/commit/f8347b1578959a00cf53318d503e46e30b25e2e6) should address the remaining issues I believe

Reply:
Thanks for this suggestion, I will check this out. 

Btw would this format all Java files?

Reply:
> This is still an open issue right? I am not sure I can find a related PR?

The issue is still open but here's the [pr](https://github.com/apache/arrow/issues/41455)

Reply:
@laurentgo I will fix this and update you later tomorrow. On a different urgent issue. Thanks for this wonderful suggestion :+1: